DOCSP-35946 write operations arrays (#2825)
jenssegers/laravel-mongodb: docs/fundamentals/write-operations.txt

example results in which it performs an insert.

The following example shows how to perform an update or insert in a single
operation by passing the ``upsert`` option to the ``update()`` method. The
example output shows the result if the match criteria passed to the
``where()`` method matches zero documents in the collection. Click the
:guilabel:`VIEW OUTPUT` button to see the document inserted by running the
code:

.. io-code-block::

   .. input:: /includes/fundamentals/write-operations/WriteOperationsTest.php
      :language: php
      :dedent:
      :start-after: begin model upsert
      :end-before: end model upsert

   .. output::
      :language: json
      :visible: false

      {
        "_id": "660c...",
        "performer": "Jon Batiste",
        "venue": "Radio City Music Hall",
        "genres": [
          "R&B",
          "soul"
        ],
        "ticketsSold": 4000,
        "updated_at": ...
      }

.. _laravel-modify-documents-arrays:

Update Arrays in a Document
---------------------------

In this section, you can see examples of the following operations that
update array values in a MongoDB document:

> - Add values to an array
> - Remove values from an array
> - Update the value of an array element

Reviewer: S: linking to each section here would be helpful
```suggestion
- :ref:`Add values to an array <laravel-modify-documents-add-array-values>`
- :ref:`Remove values from an array <laravel-modify-documents-remove-array-values>`
- :ref:`Update the value of an array element <laravel-modify-documents-update-array-values>`
```

Author: Thanks -- I was undecided about this, but if you think it's helpful, I'll add them for these examples and the ones from other PRs.

These examples modify the sample document created by the following insert
operation:

.. literalinclude:: /includes/fundamentals/write-operations/WriteOperationsTest.php
   :language: php
   :dedent:
   :start-after: begin array example document
   :end-before: end array example document

.. _laravel-modify-documents-add-array-values:

Add Values to an Array Example
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This section shows how to use the ``push()`` method to add values to an array

> in a MongoDB document. You can pass one or more values to add and set the
> optional parameter ``unique`` to ``true`` to skip adding any values

Reviewer: I: it doesn't seem accurate to say that setting `unique` to `true` skips adding "any values in the array" - maybe you could add "duplicate"?

S: I also think you can clarify this sentence a little by adding a comma:
```suggestion
in a MongoDB document. You can pass one or more values to add to the array, and you can set the
optional parameter ``unique`` to ``true`` to skip adding any duplicate values
```

Author: Good catch! I think I missed that when revising this sentence multiple times.

in the array. The following code example shows the structure of a ``push()``
method call:

.. code-block:: none
   :copyable: false

   YourModel::where(<match criteria>)
      ->push(
          <field name>,
          [<values>], // array or single value to add
          unique: true); // whether to skip existing values

> The following example shows how to add the value ``baroque`` to

Reviewer: S: add quotes
```suggestion
The following example shows how to add the value ``'baroque'`` to
```

Author: Good catch, will use double quotes to match the rest of the string values.

Author: Good catch! Started a discussion on Slack regarding whether to use single or double quotes to format string values (to address the rest of the similar comments).

the ``genres`` array field of a matching document. Click the
:guilabel:`VIEW OUTPUT` button to see the updated document:

.. io-code-block::

   .. input:: /includes/fundamentals/write-operations/WriteOperationsTest.php
      :language: php
      :dedent:
      :start-after: begin model array push
      :end-before: end model array push

   .. output::
      :language: json
      :visible: false

      {
        "_id": "660eb...",
        "performer": "Mitsuko Uchida",
        "genres": [
            "classical",
            "dance-pop",
            "baroque"
        ],
        "updated_at": ...,
        "created_at": ...
      }

.. _laravel-modify-documents-remove-array-values:

Remove Values From an Array Example
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This section shows how to use the ``pull()`` method to remove values from
an array in a MongoDB document. You can pass one or more values to remove
from the array. The following code example shows the structure of a
``pull()`` method call:

.. code-block:: none
   :copyable: false

   YourModel::where(<match criteria>)

>       ->push(

Reviewer: I: I think this should be `pull()`
```suggestion
      ->pull(
```

          <field name>,
          [<values>]); // array or single value to remove

> The following example shows how to remove array values ``"classical"`` and
> ``"dance-pop"`` from the ``genres`` array field. Click the

Reviewer: S: [nit] since single quotes are used in the code:
```suggestion
The following example shows how to remove array values ``'classical'`` and
``'dance-pop'`` from the ``genres`` array field. Click the
```

      :language: php
      :dedent:
      :start-after: begin model array pull
      :end-before: end model array pull

   .. output::
      :language: json
      :visible: false

      {
        "_id": "660e...",
        "performer": "Mitsuko Uchida",
        "genres": [],
        "updated_at": ...,
        "created_at": ...
      }

.. _laravel-modify-documents-update-array-values:

Update the Value of an Array Element Example
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This section shows how to use the ``$`` positional operator to update specific
array elements in a MongoDB document. The ``$`` operator represents the first
array element that matches the query. The following code example shows the
structure of a positional operator update call on a single matching document:

.. note::

   Currently, {+odm-short+} offers this operation only on the ``DB`` facade
   and not on the Eloquent ORM.

.. code-block:: none
   :copyable: false

   DB::connection('mongodb')
      ->getCollection(<collection name>)
      ->updateOne(
          <match criteria>,
          ['$set' => ['<array field>.$' => <replacement value>]]);

> The following example shows how to replace the array value ``"dance-pop"``
> with ``"contemporary"`` in the ``genres`` array field. Click the

Reviewer: S: same comment about single quotes as above:
```suggestion
The following example shows how to replace the array value ``'dance-pop'``
with ``'contemporary'`` in the ``genres`` array field. Click the
```
